core/storage/btree.rs

```diff
@@ -395,6 +406,202 @@ enum CursorValidState {
+// These functions below exist to avoid problems with the borrow checker
+impl CursorSeekState {
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn set_nearest_matching_cell(&mut self, matching_cell: Option<usize>) {
+        let CursorSeekState::Seeking {
+            nearest_matching_cell,
+            ..
+        } = self
+        else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *nearest_matching_cell = matching_cell;
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn get_nearest_matching_cell(&mut self) -> Option<usize> {
+        let CursorSeekState::Seeking {
+            nearest_matching_cell,
+            ..
+        } = self
+        else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        nearest_matching_cell.clone()
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn set_max(&mut self, max: isize) {
+        let CursorSeekState::Seeking { max: max_state, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *max_state = max;
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn get_max(&mut self) -> isize {
+        let CursorSeekState::Seeking { max, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *max
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn set_min(&mut self, min: isize) {
+        let CursorSeekState::Seeking { min: min_state, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *min_state = min;
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn get_min(&mut self) -> isize {
+        let CursorSeekState::Seeking { min, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *min
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn set_not_found_leaf(&mut self, not_found: bool) {
+        let CursorSeekState::Seeking { not_found_leaf, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *not_found_leaf = not_found;
+    }
+
+    /// # Safety
+    /// method can only be called on CursorSeekState::Seeking
+    fn get_not_found_leaf(&mut self) -> bool {
+        let CursorSeekState::Seeking { not_found_leaf, .. } = self else {
+            unreachable!("method can only be called on CursorSeekState::Seeking")
+        };
+        *not_found_leaf
+    }
+}
+
+#[derive(Debug)]
+/// State used for seeking
+enum CursorMoveToState {
+    Start,
+    ContinueLoop,
```

Comment:
The `ContinueLoop` state seems unnecessary since the only place where it's matched against is

```rust
            if matches!(
                self.move_to_state,
                CursorMoveToState::Start | CursorMoveToState::ContinueLoop
            )
```

and that matches `Start` too